### ISSUE TITLE
All hardware is working?

### ISSUE THREAD

**Issue by jchjian** [andylol/us9230e1#1]
I also do a similar configuration of mobile phones CM11. The CM11 you porting if audio wifi and bluetooth can be working?

**Comment by andylol** [comment]
No   not yet  but rom does boot. I guess init changes are needed   plus kernel source is closed by yusun. Our devices are clone .

**Comment by andylol** [comment]
Try using generic audio =true and setting alsa off

**Comment by jchjian** [comment]
BOARD_EGL_NEEDS_LEGACY_FB := true  set this configuration value it can boot into desktop . The audio issue i had tried a lot of different configuration in Boardconfig.mk, but it isn't working now. It had to make the CM11 once again after changed configuration in Boardconfig.mk every time, will cost about two hours each time.

**Comment by jchjian** [comment]
I will try what you just said

**Comment by andylol** [comment]
Yeah and tell me.if.it.works . And abt the wifi.and.bluetooth  they come in settings but they won't start. Logcat says cannot access wpa supplicant.  So I am thinking of.making changes in init files . When i'll get free.i'll share files with you  :)

**Comment by jchjian** [comment]
You compile it so fast? I do not know how to build the corresponding module, so i will make clean and brunch again.

**Comment by andylol** [comment]
No   I have previous log bro .. It.takes me 2 hrs. I have prebuilt chromium

**Comment by andylol** [comment]
Brother we even need se linux and audit support. But our kernel doesn't support it.if you trying yournew rom mind sharing logcat and dmesg

**Comment by jchjian** [comment]
How can i upload the files? I not add audit service into init.rc like you. Now audio wifi bluetooth can't work. Bluetooth i had tried but didn't work . So i just want to let sound go first. About sound issue if sometimes use audio_policy.tegra.so it will crash with different configuration.

**Comment by jchjian** [comment]
How can i upload the files? I not add audit service into init.rc like you. Now audio wifi bluetooth can't work. Bluetooth i had tried but didn't work . So i just want to let sound go first. About sound issue if sometimes use audio_policy.tegra.so it will crash with different configuration.

**Comment by andylol** [comment]
just paste the log in your github// make a new file in your tree and paste it and send me the link or else use paste bin :)

**Comment by andylol** [comment]
i have made a little changes in the ramdisk files and as well as board config.. No need to compile whole thing again. Just edit the ramdisk in your kitkat kernel. Hope it works. I prefer removing audio_policy.tegra from build. it causes crashes

**Comment by jchjian** [comment]
This is the log files, https://github.com/jchjian/log_files . I did't delete audio_policy.tegra in vendor-blobs, just when it crashes and i do it. As if the dev/tihci does not apply to our phones, we need the driver to create dev/hci_tty ?

**Comment by andylol** [comment]
do u mind sharing me your device tree ?
as per ti/wpan .. the new libbt vendor uses dev/hci_tty/ i havent tried it yet.

**Comment by andylol** [comment]
OK first. i checked the logs/ i'll post a file regarding what all changes i think is needed and the wifi changes i made cannot be used.. i'll write down a file. and share it with you. any place i can chat with your like facebook or somewhere ?

**Comment by andylol** [comment]
https://github.com/andylol/Logs
check this brother

**Comment by jchjian** [comment]
Bro. I can't access facebook, google, too. I am in China. We are in different time zone, maybe I'm in bed, you're at work. I just use tencent qq or wechat to chat with each other. I had uploaded my device tree, https://github.com/jchjian/android_device_Yusun_us9230. Thank you for your help.

**Comment by andylol** [comment]
Thank you for your tree.. I am able to fix the radio... Just copy /sbin /modembackup.. I am making a new build. Let's see if audio works

**Comment by jchjian** [comment]
I had tried to make audio working close for 10 days but always failed.  It had been troubling me. I don't know how can other developers fix audio issues when they are porting CM rom base on android4.1 stock rom that is not open source. I thought about seeking help on forum.xda-developers.com. But i am not good at english and sometimes can't access this url.

**Comment by andylol** [comment]
Most of the devices have kernel source which has all necessary. Drivers.. My build is complete. I'll flash it and let you know.. WiFi won't work due to WiFi c error.

**Comment by andylol** [comment]
Hello. check this out :) hope this will help us.
https://github.com/dolorespark/android_patches_hisense_m470bsa/blob/aosp444_r2/hardware/libhardware/audio-policy.diff

**Comment by jchjian** [comment]
I check this file in CM code. It like this :
#ifndef ICS_AUDIO_BLOB
#ifndef MR1_AUDIO_BLOB
    bool (*is_stream_active_remotely)(const struct audio_policy *pol,
            audio_stream_type_t stream,
            uint32_t in_past_ms);
#endif

    bool (*is_source_active)(const struct audio_policy *pol,
            audio_source_t source);
#endif
It means that we must define -DICS_AUDIO_BLOB in BoardConfig, i had tried but it compiled with a lot of error.

**Comment by jchjian** [comment]
I had checked the CM10.1 device tree of HTC ONE X. It bases on android4.1.1 and defines -DICS_AUDIO_BLOB in BoardConfig.

**Comment by jchjian** [comment]
It is too late now, i will try it tomorrow. Thank you very much :+1:

**Comment by andylol** [comment]
OK cool. We even need charger suspend support. till then i will make a common commit of patches which we both can use.
and is your storage mounting properly ?

**Comment by jchjian** [comment]
yes, the path you can see it is on init of my device tree init.us9230.rc, and then i make class_start late_start to let 'service sdcard /system/bin/sdcard -u 1023 -g 1023 -l /data/media /mnt/shell/emulated' working.

**Comment by andylol** [comment]
I/OMXCodec( 129): Codec outputs at different sampling rate than what the input stream contains (contains data at 48000 Hz, codec outputs 44100 Hz)
so i guess soln is here:
https://gist.github.com/pawitp/7945188
Does your ril work ?
Mine works.

**Comment by andylol** [comment]
i found soln to that audio policy. i have found a patch from sero7. i'll add it soon

**Comment by jchjian** [comment]
I had maked a phone call, it could be put through. So i think it can work. Does the audio work now?

**Comment by andylol** [comment]
i am compiling with it.. check out this. (hardware/libhardware)
https://github.com/andylol/Patches-for-Yusun-Tegra-devices

**Comment by jchjian** [comment]
I had modified audio_policy.h in accordance with the diff file you share and been compiling now.

**Comment by andylol** [comment]
ok ,.. i am out of space :D :P  hehe.. need to clear some packages and other stuff :D :+1:

**Comment by jchjian** [comment]
The problem was still
W/AudioPolicyManagerBase(  558): loadOutput() invalid supported devices
W/AudioPolicyManagerBase(  558): loadInput() invalid supported devices

**Comment by andylol** [comment]
plz send me your files .. quick
libandroid_servers.so
libsensorservice.so
i ened them plz upload them on git quick

**Comment by andylol** [comment]
it guess we need to make changes in audio_policy.conf. we should check out other device tree's of tegra variant. i think it has to do something with it.

**Comment by andylol** [comment]
i am getting same error as yours. just my libsensors. something problem.
Next we come to bluetooth .add this and try
https://github.com/andylol/Patches-for-Yusun-Tegra-devices/blob/master/bluetooth_fix

**Comment by andylol** [comment]
i checked out hardware/libhardware/audio/audiopolicymanager.cpp . it uses audio_policy as base of
emulator. so i am going to check grouperz audio policy

**Comment by andylol** [comment]
share me your logcat. new one.

**Comment by jchjian** [comment]
the patch of bluetooth you share i will try it, i just send you a email, do you receive it. libbt-vendor.so will open the dev/hci_tty device, but we haven't, if it effects?

**Comment by andylol** [comment]
we are just giving it a try. we still dont know about it. just give it a try. and is the camera working for you ? . since i am still in boot loop i cant test it. but you can. i guess i have added few lines which are not necessary. so sensors are giving trouble. Next comes wifi.. it isnt loading the bin file.
"wl12xx: ERROR could not get nvs file wl1271-nvs.bin: -2"
so in init we have to locate it by adding its somethng like  /mnt/ or where ever the file is. leave that to me :D .. just test the bluetooth and see.

**Comment by andylol** [comment]
is there any way to decompile .so files so that we can get help from thm :P :D

**Comment by jchjian** [comment]
http://processors.wiki.ti.com/index.php/TI-Android-JB-PortingGuide here we can see how ti wifi and bluetooth work basic. I think it will help us. we can decompile .so files like : 
prebuilt/linux-x86/toolchain/arm-eabi-4.4.3/arm-eabi-objdump -S mylib.so > mylib.asm . asm file is assembler source file.

**Comment by jchjian** [comment]
your rom still can't boot into desktop?

**Comment by andylol** [comment]
did the bluetooth thing work.. my rom does boot. last build i added an patch unknowingly :D so i have to remove and compile it again. Should we use htc onex audio.h . i'll try it tomorrow.. Okay i'll have a look at that tree. our logs are just same :D . i guess we need to change audio policy conf

**Comment by jchjian** [comment]
i had used htc onex audio.h before, but it didn't work, and i hadn't checked logcat just passed it. i don't know if there some help for us, we may have a try.

**Comment by andylol** [comment]
ok check this line.
The audio HAL uses the file mixer_paths.xml to configure the audio driver
we have added the wifi path in board conf which dis just reference. if u check actual lib it must be lib cmd wl12xx sdio  . but if we set it to this. we get wifi c error. and to fix wifi c error we need libnl files.

**Comment by andylol** [comment]
hey you there. i have made changes in my tree as per the link. i will upload a new and final tree. which we can use a base independently :)

**Comment by jchjian** [comment]
my logs of dumpstate, https://github.com/jchjian/log_files/blob/master/dump.log. i had changed /dev/hci_tty to /dev/ttyHS2 in hardware/ti/wpan/bluedroid_wilink/libbt-vendor-ti.c just now, but it did't work, too.

**Comment by jchjian** [comment]
everyday face of these problems and keep trying time and time again, problems was still, and i have a little impatient, heart is broken, haha.

**Comment by andylol** [comment]
audio.. i dont know what we can do now.. did u try pushing a2dp.default.so from stock to rom. my rild isnt working .. i dunno why.. plus check this out. nor is my internal storage/ wl127bin not found.. hmmm got to do something of it 
https://github.com/randomblame/android_device_hisense_m470/blob/taktik/bluetooth/bt_vendor.conf

**Comment by andylol** [comment]
can u give this a try. ? 
BOARD_USES_AUDIO_LEGACY=true
Then it will use cm audio files. i am making build clean. Its 8PM here now. i am making few more changes in my tree and checking

**Comment by jchjian** [comment]
BOARD_USES_AUDIO_LEGACY=true i had tried, didn't work too, if we need to find some tree have audio open source codes similar like us? wl127bin will no error if we use wl12xx*.ko files in modules not compat

**Comment by andylol** [comment]
I have modified my tree again . will keep trying. now kept for brunching. will let u know the results
Took audio and camera wrappers from p880. even libstagefright. healthd for our batterystats.
i had added vendor_bt. which defines ttyhs2 as bluetooth device and many more changes. if u want i'll upload my tree

**Comment by jchjian** [comment]
Ok, upload files you said,  thank you

**Comment by andylol** [comment]
here it is 
https://github.com/andylol/android_device_xolo_us9230e1

**Comment by andylol** [comment]
brother do u have any idea abt this. i copied the audioflinger.so and audio policy tegra from ktouch rom
I/ServiceManager( 2876): Waiting for service media.nvidia.audio_alsa...
I/ServiceManager( 2380): Waiting for service media.audio_policy...
I/ServiceManager( 2876): Waiting for service media.nvidia.audio_alsa...

**Comment by jchjian** [comment]
if system/bin/servicemanager about this? also some other so files associated

**Comment by andylol** [comment]
i tried pushing most of them i know..  the audio blobs are perfect for our device . we just need littlebit of mixes .

**Comment by jchjian** [comment]
yes, i had used the CM11 ROM of ktouch v9 on my phone, and receiver sound normal, but speaker sound only a second when playing media file.

**Comment by jchjian** [comment]
it just all to use ktouch v9 ROM except boot.img

**Comment by jchjian** [comment]
if we try to use some audio files of  ktouch v9 CM ROM, i think audio of our phone will close to normal

**Comment by andylol** [comment]
yes it might work. just take a logcat we will come to know about it. IS camera working ? Can u just upload your rom with ktouch audio files. i will download it so that we both can sync together

**Comment by jchjian** [comment]
I just returned back to my computer. I will try to use audio files of ktouch ROM, if it work i upload it to github.

**Comment by andylol** [comment]
Can u provide me link to the ktouch rom. i'll download it
ktouch v8 uses broadcom while we use ti. so bluetooth wont work .

**Comment by jchjian** [comment]
CM11 ktouch V8 and v9 4.4.4 ：  link: http://pan.baidu.com/s/1c0nJJkk  password: ofmu

**Comment by andylol** [comment]
okay v9 is ap33h . same as ours. i have ktouch v8 miui rom 4.4.4

**Comment by andylol** [comment]
i have flashed the rom.. systemui crashing. can u help me with it. it causes no adb. what all files i have to edit or add from our rom can u tell me ?

**Comment by jchjian** [comment]
we just need to use most of our proprietary files to replace ktouch's rom, and some need to use themself, i think. How do you port?

**Comment by andylol** [comment]
i flashed my kitkat rom/ i am trying the audio libs .we need to deodexed the framework and then push it. as per this
https://thenewcircle.com/s/post/1057/MarakanaAndroidServicesBlackMagic.pdf

**Comment by andylol** [comment]
any progress my friend ?

**Comment by jchjian** [comment]
if all use ktouch's audio files, just receiver sound, if use some it will crash when open settings-sound and not a sound. what the annoying problems, i was about to give up.

**Comment by jchjian** [comment]
Bro, my modem partition has been formatted, may you dump a share for me, thanks

**Comment by jchjian** [comment]
can you push all the files in /mnt/modem/data/factory and upload, thank you very much

**Comment by jchjian** [comment]
i just know a decompilation software called ida pro with arm decompiler from intelnet, it can convert the binary program into source code, like android files, .so .a .o framework app, and so on. i don't konw if it's real.

**Comment by andylol** [comment]
I can't give you that .it contains my imei. I have modem.IMG file.  Should I upload it. 
 But that will not restore your network. you need the imei.bin and product config of your phonr

**Comment by jchjian** [comment]
Other people have given it me, now my network back to normal. Thank you all the same. cm11 still several issues still not resolved.

**Comment by andylol** [comment]
brother can u upload your ktouch rom which works in your phone. then i might be able to help. since i am not getting the right mix of blobs
